src/core/server.cc

```diff
@@ -294,150 +293,354 @@ class HTTPServiceImpl {
+    evhtp_bind_socket(htp_, "0.0.0.0", port, 1024);
+    worker_ = std::thread(event_base_loop, evbase_, 0);
+    return tensorflow::Status::OK();
+  }
+
+  return tensorflow::Status(
+      tensorflow::error::ALREADY_EXISTS, "HTTP server is already running.");
+}
+
+tensorflow::Status
+HTTPServiceImpl::Stop()
+{
+  if (worker_.joinable()) {
+    event_base_loopexit(evbase_, NULL);
+    worker_.join();
+    return tensorflow::Status::OK();
+  }
+
+  return tensorflow::Status(
+      tensorflow::error::UNAVAILABLE, "HTTP server is not running.");
+}
+
+std::string
+HTTPServiceImpl::ToMethodString(htp_method method)
 {
-  return
-      [this](tfs::net_http::ServerRequestInterface* req) { this->Handle(req); };
+  switch (method) {
+    case htp_method_GET:
+      return "GET";
+    case htp_method_POST:
+      return "POST";
+    case htp_method_PUT:
+      return "PUT";
+    case htp_method_DELETE:
+      return "DELETE";
+    default:
+      return "UNKNOWN: " + std::to_string(method);
+  }
 }
 
 void
-HTTPServiceImpl::Handle(tfs::net_http::ServerRequestInterface* req)
+HTTPServiceImpl::Dispatch(evhtp_request_t* req, void* arg)
 {
-  LOG_VERBOSE(1) << "HTTP request: " << req->http_method() << " "
-                 << req->uri_path();
+  (static_cast<HTTPServiceImpl*>(arg))->Handle(req);
+}
 
-  tfs::net_http::HTTPStatusCode status =
-      tfs::net_http::HTTPStatusCode::BAD_REQUEST;
+void
+HTTPServiceImpl::Handle(evhtp_request_t* req)
+{
+  LOG_VERBOSE(1) << "HTTP request: " << ToMethodString(req->method) << " "
+                 << req->uri->path->full;
 
   std::string endpoint, rest;
   if (RE2::FullMatch(
-          std::string(req->uri_path()), api_regex_, &endpoint, &rest)) {
+          std::string(req->uri->path->full), api_regex_, &endpoint, &rest)) {
     // health
     if (endpoint == "health") {
-      status = Health(req, rest);
+      Health(req, rest);
     }
     // profile
     else if (endpoint == "profile") {
-      status = Profile(req, rest);
+      Profile(req, rest);
     }
     // infer
     else if (endpoint == "infer") {
-      status = Infer(req, rest);
+      Infer(req, rest);
     }
     // status
     else if (endpoint == "status") {
-      status = Status(req, rest);
+      Status(req, rest);
     }
   }
-
-  if (status != tfs::net_http::HTTPStatusCode::OK) {
-    LOG_VERBOSE(1) << "HTTP error: " << req->http_method() << " "
-                   << req->uri_path() << " - " << static_cast<int>(status);
-  }
-
-  req->ReplyWithStatus(status);
 }
 
-tfs::net_http::HTTPStatusCode
-HTTPServiceImpl::Health(
-    tfs::net_http::ServerRequestInterface* req, const std::string& health_uri)
+void
+HTTPServiceImpl::Health(evhtp_request_t* req, const std::string& health_uri)
 {
   ServerStatTimerScoped timer(
       server_->StatusManager(), ServerStatTimerScoped::Kind::HEALTH);
 
-  if (req->http_method() != "GET") {
-    return tfs::net_http::HTTPStatusCode::METHOD_NA;
+  if (ToMethodString(req->method) != "GET") {
```

Comment:
Converting to string is inefficient. Why not directly compare to htp_method_GET?